libdfu/dfu-device.c

```diff
@@ -1363,15 +1368,17 @@ dfu_device_open (DfuDevice *device, DfuDeviceOpenFlags flags,
 			     priv->platform_id);
 		return FALSE;
 	}
+	if (priv->dev_locker != NULL) {
+		g_debug ("device already open, ignoring");
+		return TRUE;
+	}
 
 	/* open */
-	if (!g_usb_device_open (priv->dev, &error_local)) {
-		if (g_error_matches (error_local,
-				     G_USB_DEVICE_ERROR,
-				     G_USB_DEVICE_ERROR_ALREADY_OPEN)) {
-			g_debug ("device already open, ignoring");
-			return TRUE;
-		}
+	locker = fu_device_locker_new (priv->dev,
+				       (FuDeviceLockerFunc) g_usb_device_open,
+				       (FuDeviceLockerFunc) g_usb_device_close,
+				       &error_local);
```

Comment:
Do you envision that realistically anything but USB will need this?  Rather than every instance needing to define the open and close function why not just make that implicit?

Reply:
Really good question. I did originally put this object in libgusb as a helper, but I can see it useful to chain up the "parent" open and close methods too, e.g. `dfu_device_open()`, `fu_device_altos_tty_open()`, `fu_device_ebitdo_open()` and `synapticsmst_device_open()` -- I wanted a sanity check before I put in the hours converting the 2nd layer. I can do this if you think it'd be useful.

Reply:
OK then yeah I think this approach makes sense to me.  As you've seen in the past with unifying, there can be weird problems with keeping devices open so it does leave flexibility in the future.

Reply:
I've added some more different device types to the latest commit - I've managed to break DFU in the process, so it's WIP.